railgun/manager/automagic_controller_enablement.go

```diff
@@ -0,0 +1,52 @@
+package manager
+
+import (
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	knativev1alpha1 "knative.dev/networking/pkg/apis/networking/v1alpha1"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	"github.com/kong/kubernetes-ingress-controller/pkg/util"
+	konghqcomv1 "github.com/kong/kubernetes-ingress-controller/railgun/apis/configuration/v1"
+	"github.com/kong/kubernetes-ingress-controller/railgun/internal/ctrlutils"
+	"github.com/kong/kubernetes-ingress-controller/railgun/pkg/config"
+)
+
+// autoCTRL will automatically enable or disable controllers based on whether they've been
+// explicitly disabled (e.g. util.EnablementStatusDisabled) and whether or not the API for
+// that controller is available and loaded in the Kubernetes API.
+func autoCTRL(mgr manager.Manager, log logr.Logger, c *config.Config) {
```

Comment:
```suggestion
func autoCTRL(mgr manager.Manager, log logr.Logger, c *config.Config) {
Don't see any difference from previous function FlipKnativeController
For generic, ctrl name, GVR, ctrlenablement should be passed in as parameter rather than specific GVR within function.

Suggestion:  use the PR resolving the reported bug rather than generic/refactor.
```

Reply:
> Don't see difference func autoCTRL from previous function FlipKnativeController

If this PR is accepted, `FlipKnativeController` would be _removed_. The `autoCTRL` function is not intended as a replacement for this, it replaces the logic that would _append to the controller list_.

> For generic, ctrl name, GVR, ctrlenablement should be passed in as parameter rather than specific GVR within function.

I'm not sure I understand: this function passes in the enablement configurations via the `*config.Config` object, does this not align with what you're looking for?

> Please use the PR resolving the reported bug rather than generic/refactor.

I think you mean to say that this PR should _only_ focus on fixing the originally reported bug in the most minimal way possible, is that correct?

Reply:
No. the controller configuration only defined how controller would be up specifically.
i.e pass in gvr( schema.GroupVersionResource) as a parameter and then use generic function client.RESTMapper().KindFor(gvr) do the check.

Reply:
> To be more generic, one way would be to pass in gvr( schema.GroupVersionResource) as a parameter and then use generic function client.RESTMapper().KindFor(gvr) do the check.

Ah, I understand. I don't think passing in the GVRs is necessary: as a private function this is encapsulated and specifically there to assist the `manager.Run` function. I submit that it's legitimate to have the functionality here be specific and static since it's a private function meant only for consumption under a very specific premise, could you be persuaded in that regard?

Reply:
What I mean need understand/clarify how the exact issue is fixed from the change.

Reply:
> What I mean need understand/clarify how the exact issue is fixed from the PR change (refactor code could not fix bug).

The `FlipKnativeController` functionality has been completely removed, which fixes the bug reported in #1466. However, this PR is _not just about fixing that bug_ it's also about some updates to improve code maintenance. This multiple-agenda is covered in the original PR description:

```
This cleans up some of the dynamic controller setup by taking pre-processing setups to flag them as explicitly enabled or disabled based on their current util.EnablementStatus and API availability.

This also fixes a bug (see #1466) wherein the controller manager could not start properly due to a hard dependency on knative.Ingress, even when the controller was explicitly disabled.
```

Is there some way I should improve that description?

Reply:
(continuing this conversation in slack and maybe zoom)